### PR TITLE
Impact definitions as a record

### PR DESCRIPTION
**Pull request by magopian**
This is a huge one ... the aim is to get rid of the "dynamic" `impacts.json` and have it in an elm file instead. This then doesn't need all the loading and parsing, and should give us far less "passing the impacts around", and less dealing with results and maybes.

### CODE REVIEW

File: src/Data/Impact.elm [hunk 1]
```diff
@@ -150,29 +101,14 @@ type alias ProtectionAreas =
...
-    , source = { label = "N/A", url = "https://example.com/" }
-    , label = "Not applicable"
-    , description = "Not applicable"
-    , unit = "N/A"
-    , decimals = 0
-    , quality = GoodQuality
-    , pefData = Nothing
-    , ecoscoreData = Nothing
-    , scopes = []
-    }
```
Comment (n1k0) [inline src/Data/Impact.elm:165]: Thanks for that 🙏

File: src/Data/Impact.elm [hunk 2]
```diff
@@ -340,15 +172,15 @@ type Impacts
...
 noImpacts : Impacts
 noImpacts =
-    AnyDict.empty toString
+    AnyDict.empty Definition.toString
         |> Impacts
 
 
-impactsFromDefinitons : List Definition -> Impacts
+impactsFromDefinitons : Impacts
 impactsFromDefinitons =
```
Comment (n1k0) [inline src/Data/Impact.elm:180]: This fn might want to be renamed `init` or `empty`.

File: src/Data/Impact.elm [hunk 3]
```diff
@@ -527,16 +355,8 @@ computeAggregatedScore getter defs (Impacts impacts) =
 
 parseTrigram : Scope -> Parser (Trigram -> a) a
 parseTrigram scope =
-    let
-        trigrams =
-            -- FIXME: find a way to have this check performed automatically from impacts db
-            "acd,bvi,cch,ecs,etf,etf-c,fru,fwe,htc,htc-c,htn,htc-c,ior,ldu,mru,ozd,pco,pef,pma,swe,tre,wtu"
-                |> String.split ","
-    in
```
Comment (n1k0) [inline src/Data/Impact.elm:535]: Insert tears of joy emoji here.

**Review by n1k0 — APPROVED** [review]
Looks fantastic mate, well done 🤩

**Comment by paulboosz** [comment]
Just my 2 scents here, I find impacts.json useful for sanity checks : 
It's the only place where I can find the normalization and weighting to compute the PEF score and "score d'impacts" in a computer readable format. So using the API and impacts.json I can recompute the differents score using the raw impacts. And if someone updates the weighting I don't have to update my script

**Comment by magopian** [comment]
So, here's what I propose:
1/ keep the impacts.json file as the single source of truth
2/ investigate ways to automatically convert it to the Definition.elm we need for this PR to land

**Comment by magopian** [comment]
So this PR has pivoted to automatically generate the `src/Data/Impact/Definition.elm` file from the `public/data/impacts.json` file which is the single source of truth regarding impacts.

It uses [elm-codegen](https://package.elm-lang.org/packages/mdgriffith/elm-codegen/latest/), and the generation script is in `codegen/Generate.elm`.

On each run of the ui/server/tests, re-generate the file to make sure we have the most up to date definitions.

**Comment by magopian** [comment]
And yet another pivot 😅 : no more code generation, in this very case it seems it was adding more complexity than it was solving. Maybe in the future for a better suited use case!

Not having the code generation means we'll need to manually add new impacts to the `Definition.elm` file, if the case ever occurs (it's very rare, so i'm confident it won't be that much of a task).

**Review by n1k0 — APPROVED** [review]
Looks totally reasonable 👍

File: src/Data/Impact/Definition.elm [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    | Pef
+    | Pma
+    | Swe
+    | Tre
+    | Wtu
+
+
+type alias Definition =
+    { trigramString : String
+    , trigram : Trigram
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Data/Impact/Definition.elm:60]: I believe it's a leftover from the "code generation and impacts.json as static data" need, it's not needed anymore, i'll just remove it 🙏

File: src/Data/Impact/Definition.elm [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+        Wtu ->
+            "wtu"
+
+
+toTrigram : String -> Result String Trigram
+toTrigram str =
+    case str of
+        "acd" ->
+            Result.Ok Acd
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Data/Impact/Definition.elm:274]: Same here, it's a code generation artifact that's no longer needed 👍

File: src/Page/Food/Builder.elm [hunk 6]
```diff
@@ -1385,7 +1390,7 @@ impactTabsView db model results =
             [ case model.activeImpactsTab of
                 DetailedImpactsTab ->
                     results.total
-                        |> Impact.getAggregatedScoreData db.impacts .ecoscoreData
+                        |> Impact.getAggregatedScoreData model.db.impactDefinitions .ecoscoreData
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Page/Food/Builder.elm:1393]: Actually, we have a loaded food db in the model (not a WebData like in the session), so I'll make sure we use this every where in this file 👍

**Pull request merged**